articles/libraries/auth0-android/save-and-refresh-tokens.md

```diff
@@ -0,0 +1,117 @@
+
+First, add the library dependency to your build.gradle file:
+
+```gradle
+compile 'com.auth0.android:auth0:1.+'
+```
+
+Next, decide which class to use depending on your Android SDK target version:
+- If you require APIs lower than 21 (Lollipop) create a new instance of `CredentialsManager`. This class makes use of the `Context.MODE_PRIVATE` to store the credentials in a key-value preferences file, accessible only to your app. The data is stored in plaintext.
+- If you require APIs equal or greater than 21 (Lollipop) create a new instance of `SecureCredentialsManager`. This class has the same methods as the one above, but encrypts the data before storing it using a combination of _RSA_ and _AES_ algorithms along with the use of [Android KeyStore](https://developer.android.com/reference/java/security/KeyStore.html).
+
+
+## Using the CredentialsManager class
+
+### Setup
+Create a new instance by passing an `AuthenticationAPIClient` and a `Storage` implementation.
+
+```java
+Auth0 auth0 = new Auth0(this);
+AuthenticationAPIClient apiClient = new AuthenticationAPIClient(auth0);
+CredentialsManager manager = new CredentialsManager(apiClient, new SharedPreferencesStorage(this));
+```
+
+### Current State of the Authentication
+By asking for credentials existence upon launch and clearing them on logout, the manager can be used to manage the state of the authentication inside your app.
+
+Stored credentials are considered valid if they have not expired or can be refreshed. Check if a user has already logged in:
+
+```java
+boolean loggedIn = manager.hasValidCredentials();
+```
+
+When you want to log out the user from your app, remove the stored credentials and direct them to the login screen.
+
+```java
+manager.clearCredentials();
+//Show login screen
+```
+
+### Obtain Credentials
+Because the credentials may need to be refreshed against Auth0 Servers, this method is a-synchronic. Pass a callback implementation where you'd like to receive the credentials. Credentials returned by this method upon success are always valid.
+
+```java
+manager.getCredentials(new BaseCallback<Credentials, CredentialsManagerException>() {
+    @Override
+    public void onSuccess(Credentials credentials) {
+        //Use credentials
+    }
+
+    @Override
+    public void onFailure(CredentialsManagerException error) {
+        //No credentials were previously saved or they couldn't be refreshed
+    }
+});
+```
+
+::: note
+If the credentials were renewed, the manager will automatically save the new value for you.
+:::
+
+### Save new Credentials
+You can save the credentials obtained upon log in in the manager. If they contain a valid `refresh_token` they will be refreshed automatically if they expire, in other case, they will be deemed invalid.
```

Comment:
Don't think you need to talk about refresh token here, the section is about 'saving'.  The refresh stuff can be covered in the previous note? Can mention it also saves it automatically (assuming it does)

Reply:
I think it's the other way around. We should talk about the refresh logic in the "retrieve" rather than in the save. Save doesn't do refresh, and will accept credentials without refresh_token just fine. I'll add in the previous step something to indicate that we store the refreshed credentials.